### ISSUE TITLE
Drop CRDs manifest but install CRDs by velero install

### ISSUE THREAD

**Issue by jenting** [vmware-tanzu/helm-charts#44]
Right now, the velero helm chart CRD manifest is not in-sync with velero repo.
Also the command `velero install --crds-only` supports install CRDs.

I'm wondering that is it better to drop helm-chart CRDs manifest but install the CRDs by `veleror install --crds-only`?
This way makes maintaining the CRDs from the single source and does not need to address sync CRD manifest from velero repo to velero helm chart.

Are there any other cons I did not think about? Thanks.

**Comment by carlisia** [comment]
@cpanato do you have any opinion?

I'm going to add this to next's week community meeting agenda.

**Comment by carlisia** [comment]
https://hackmd.io/Jq6F5zqZR7S80CeDWUklkA

**Comment by cpanato** [comment]
I saw other projects that don't install the CRDs as part of the helm installation, but I don't recall which ones were :(

I think this is a good idea but we need to make sure we output the additional commands we need to apply after we install the chart.

**Comment by jenting** [comment]
I was work on this previously, if this is a good approach, I could help on this 😉

**Comment by nrb** [comment]
I've also seen other projects do this, due to the way Helm's designed.

It comes down to Helm wanting to manage a release's lifecycle, while a CRD by definition must live much longer than any one release. Helm has tried to fix it in several ways, and so far I haven't seen any of them be particularly reliable in Helm 2; I've not done any experimenting in Helm 3 to know if the situation has improved. 

I do agree that maintaining one source of definitions would be good, because having them get out of sync is not good for anyone.

**Comment by carlisia** [comment]
Just something to think about: we have the intention of dropping the `velero install` command: https://github.com/vmware-tanzu/velero/pull/2202

**Comment by jenting** [comment]
My idea on how to achieve this is:
1. Run a Job which executes "velero install --crds-only" with helm pre-install hook to make sure that the CRDs are installed.
2. The tests resources are installed with helm post-install hook.

However, after testing on helm3, it seems like the helm would not waits the Job executes completed then install the other resources. It means that right now I cannot find a way to achieve this.

BTW, I found two problems in current helm charts:
1. The CRDs are out-of-sync compares to current Velero version v1.4.2
2. The helm would not handles update existing CRDs when helm upgrade. Ref to https://helm.sh/docs/chart_best_practices/custom_resource_definitions/

**Comment by carlisia** [comment]
We'd need to generate the CRDs originally already containing these annotations right:

https://github.com/vmware-tanzu/helm-charts/blob/main/charts/velero/crds/backups.yaml#L8-L9

**Comment by jenting** [comment]
> We'd need to generate the CRDs originally already containing these annotations right:
> 
> https://github.com/vmware-tanzu/helm-charts/blob/main/charts/velero/crds/backups.yaml#L8-L9

According to discussion https://github.com/vmware-tanzu/velero/pull/2779, we wouldn't generate the CRDs with the helm annotation included.

I have an idea that we could have a CI check (by script) to verify the current helm chart app version is aligned with vmware-tanzu/velero released version CRDs and the script also helps to add annotation/labels to the CRDs.

**Comment by caiobegotti** [comment]
What's the current supported and correct way of instaling the Helm chart with the CRDs? Right now if I render the chart I get no CRDs with a `helm template` (this way I can store it in my Git repository). Does it only installs CRDs if I use `helm install`? I tried installing the CRDs with the Velero CLI as suggested here but that's very sub-optimal and I can'get quite get it to work with Kustomize as I'm also trying to keep it under version control.

So, TL;DR: I'm lost about how 's the correct way to get CRDs ready before deploying Velero with Helm (templated).

**Comment by jenting** [comment]
@caiobegotti I assume you're using helm3, please add `--include-crds` flag to display CRDs in the templated output.

**Comment by jenting** [comment]
Close the PR because we wrapper the way to install/upgrade CRDs by the `velero install --crds-only --dry-run -oyaml`, this makes `helm template` display the CRDs manifest not work anymore.

Open a new issue to address the `helm upgrade` does not upgrade the CRDs manifest, close it.
